### PR TITLE
[FGJ-58] Ability to build a gateway without a Wallet

### PR DESCRIPTION
**Pull request by mrudav-shukla**
**Type of change:**

- New Feature

**Description:**

- This feature adds the ability to build a Gateway without using an identity stored in the Wallet.

- Gateway.Builder now has an overload of identity() method which supports this feature.

- Currently, only X509Identity is supported for this.

**JIRA Ticket**

- [FGJ-58](https://jira.hyperledger.org/browse/FGJ-58)

### CODE REVIEW

File: src/test/resources/scenario/FabCar.feature [hunk 1]
```diff
@@ -29,6 +29,26 @@ Feature: Configure Fabric using SDK and submit/evaluate using a network Gateway
 		    }
 		    """
 
+	Scenario: Using a Gateway without Wallet I can submit and evaluate transactions on instantiated node chaincode
+		Given I have a gateway as user User1 using the tls connection profile without a Wallet
```
Comment (bestbeforetoday) [inline src/test/resources/scenario/FabCar.feature]: Maybe `I have a gateway as identity User1 using the tls connection profile` to avoid the `without a Wallet` wording?

File: src/test/java/scenario/ScenarioSteps.java [hunk 2]
```diff
@@ -183,6 +183,18 @@ public ScenarioSteps() throws IOException {
                     }
                 });
 
+	    Given("I have a gateway as user {word} using the {word} connection profile without a Wallet",
+	            (String userName, String tlsType) -> {
```
Comment (bestbeforetoday) [inline src/test/java/scenario/ScenarioSteps.java]: Some of this code is a duplicate of the `I have a gateway as user {word} using the {word} connection profile` step definition. It would be good to pull out the common parts to another function to avoid duplication.

File: src/test/java/scenario/ScenarioSteps.java [hunk 3]
```diff
@@ -183,6 +183,18 @@ public ScenarioSteps() throws IOException {
                     }
                 });
 
+	    Given("I have a gateway as user {word} using the {word} connection profile without a Wallet",
```
Comment (bestbeforetoday) [inline src/test/java/scenario/ScenarioSteps.java]: The user name argument isn't being used. The (currently) only supported value could either be inserted directly into the stepdef text, or the value of the `userName` argument should probably be checked that it matches the expected value in the code.

File: src/test/resources/scenario/FabCar.feature [hunk 4]
```diff
@@ -29,6 +29,26 @@ Feature: Configure Fabric using SDK and submit/evaluate using a network Gateway
...
+		And I connect the gateway
+		And I use the mychannel network
+		And I use the fabcar contract
+		When I prepare a createCar transaction
+        And I submit the transaction with arguments ["CAR10", "Trabant", "601 Estate", "brown", "Simon"]
```
Comment (bestbeforetoday) [inline src/test/resources/scenario/FabCar.feature]: There is a mix of spaces and tabs here

File: src/test/resources/scenario/FabCar.feature [hunk 5]
```diff
@@ -29,6 +29,26 @@ Feature: Configure Fabric using SDK and submit/evaluate using a network Gateway
...
+		And I prepare a queryCar transaction
+        And I evaluate the transaction with arguments ["CAR10"]
```
Comment (bestbeforetoday) [inline src/test/resources/scenario/FabCar.feature]: There is a mix of space and tabs here

**Comment by bestbeforetoday** [comment]
You'll need to amend your commit using `--signoff` to include the line the DCO check is looking for. Also rebase on _master_ to make sure the PR is up-to-date.

File: src/test/java/scenario/ScenarioSteps.java (hunk 3 above)
Comment (mrudav-shukla) [reply src/test/java/scenario/ScenarioSteps.java]: Right. Can I rephrase it to "I have a gateway with identity User1 using the {word} connection profile"?

Comment (bestbeforetoday) [reply src/test/java/scenario/ScenarioSteps.java]: Yes, that looks fine. Just remember to change the accompanying function signature to just have the one (String tlsType) parameter.

File: src/test/resources/scenario/FabCar.feature (hunk 1 above)
Comment (mrudav-shukla) [reply src/test/resources/scenario/FabCar.feature]: For the scenario part, can I rephrase it to: "Using a Gateway built with X509Identity I can submit and evaluate transactions on instantiated node chaincode"?

**Comment by mrudav-shukla** [comment]
@bestbeforetoday  I tried adding the sign-off in all the recent commits. Do I also need to amend the earlier commits?

**Comment by bestbeforetoday** [comment]
As discussed by chat, all the commits need to have a valid sign-off. I usually find the easiest way to achieve this is to squash all the changes to a single commit with the sign-off line.

File: src/main/java/org/hyperledger/fabric/gateway/Gateway.java [hunk 6]
```diff
@@ -132,6 +132,14 @@ static Builder createBuilder() {
          */
         Builder identity(Wallet wallet, String id) throws IOException;
 
+        /**
+         * Specifies the identity that is to be used to connect to the network.  All operations
+         * under this gateway connection will be performed using this identity.
+         * @param identity The Identity using which operations has to be performed.
```
Comment (bestbeforetoday) [inline src/main/java/org/hyperledger/fabric/gateway/Gateway.java]: The description of the identity parameter doesn't make sense to me. Even something as simple as "An identity" would be clearer to me.

File: src/test/resources/scenario/FabCar.feature [hunk 7]
```diff
@@ -10,24 +10,44 @@ Feature: Configure Fabric using SDK and submit/evaluate using a network Gateway
...
+            }
+            """
+
+    Scenario: Using a Gateway with an X509Identity I can submit and evaluate transactions on instantiated node chaincode
+        Given I have a gateway with identity User1 using the tls connection profile
+        And I connect the gateway
+        And I use the mychannel network
+        And I use the fabcar contract
+        When I prepare a createCar transaction
+        And I submit the transaction with arguments ["CAR10", "Trabant", "601 Estate", "brown", "Simon"]
```
Comment (bestbeforetoday) [inline src/test/resources/scenario/FabCar.feature]: A different ID than `CAR10` should be used. That ID is used by the scenario test above so it is possible one test may not be working correctly but appear that it is since it is seeing state set by the other test. Alternatively, you could use the same ID but with different information so we know we're querying the correct state.
